Change motor control input from Amps to Nm

PAJohnson opened the pull request:
The purpose of this PR is to change from units of Amps to units of Nm for the input value to the motor control. This affects multiple parts of the ODrive code structure, including the communication interfaces.

PAJohnson committed:
added torque_constant to motor config struct and interface file
changed all instances of current_setpoint to torque_setpoint
modified limitVel() to use torque instead of current units
removed effective_current_lim, added effective_torque_lim()
added torque_lim to motor config struct and interface file

PAJohnson committed:
Modified docs and communication interfaces to reflect change from A to Nm for motor control input.
Renamed vel_integrator_current_ to vel_integrator_torque_
Removed input_current_ from controller, added input_torque_

madcowswe commented on Firmware/MotorControl/controller.cpp:
```diff
@@ -153,29 +153,29 @@ bool Controller::update(float* current_setpoint_output) {
-            current_setpoint_ = (step / current_meas_period) * config_.inertia;
+            torque_setpoint_ = (step / current_meas_period) * config_.inertia;
         } break;
         case INPUT_MODE_CURRENT_RAMP: {
```
We should also rename this to Torque ramp.

PAJohnson replied:
I think you meant this in reference to INPUT_MODE_CURRENT_RAMP. I changed that to INPUT_MODE_TORQUE_RAMP.

madcowswe commented on Firmware/MotorControl/motor.cpp:
```diff
@@ -177,19 +177,19 @@ bool Motor::do_checks() {
 }
 
-float Motor::effective_current_lim() {
+float Motor::effective_torque_lim() {
```
Hm actually I think all the limits involved here make more sense to be in terms of current. I would keep this function as effective_current_lim, and apply the conversion at every applicable call site.

madcowswe commented on Firmware/MotorControl/motor.hpp:
```diff
@@ -29,6 +29,7 @@ class Motor : public ODriveIntf::MotorIntf {
         float Id_measured; // [A]
         float I_measured_report_filter_k;
         float max_allowed_current; // [A]
+        float max_allowed_torque; // [Nm]
```
"max allowed current" is a hardware limit, as noted earlier, probably best to keep the current limits as current and not torque.

madcowswe commented on Firmware/communication/ascii_protocol.cpp:
```diff
@@ -108,15 +108,15 @@ void ASCII_protocol_process_line(const uint8_t* buffer, size_t len, StreamSink&
-        float pos_setpoint, vel_limit, current_lim;
-        int numscan = sscanf(cmd, "q %u %f %f %f", &motor_number, &pos_setpoint, &vel_limit, &current_lim);
+        float pos_setpoint, vel_limit, torque_lim;
+        int numscan = sscanf(cmd, "q %u %f %f %f", &motor_number, &pos_setpoint, &vel_limit, &torque_lim);
```
Here is a good example where we do in fact want a torque limit (not current limit), as it's a proxy for an acceleration limit. I guess we should have both a torque and current limit be observed seperately.

madcowswe commented on docs/getting-started.md:
```diff
@@ -346,9 +346,9 @@ Set the velocity ramp rate (acceleration): `axis.controller.config.vel_ramp_rate
+### Torque control
 Set `axis.controller.config.control_mode = CONTROL_MODE_CURRENT_CONTROL`.<br>
-You can now control the current with `axis.controller.input_current = 3` [A].
+You can now control the torque with `axis.controller.input_torque = 3` [Nm].
```
3Nm is very large (comared to 3 A on common motors). Perhaps use 0.1Nm in the example.

madcowswe commented on tools/.vscode/launch.json:
```diff
@@ -9,7 +9,7 @@
             "request": "launch",
             "stopOnEntry": true,
-            "pythonPath": "${config:python.pythonPath}",
+            "pythonPath": "${config:python.interpreterPath}",
```
Is this change intended? What is the reason it is changed (I don't know what either does)

PAJohnson replied:
This change was not intended and I also do not know what it does.

madcowswe commented on Firmware/MotorControl/motor.cpp:
```diff
@@ -440,13 +440,14 @@ bool Motor::FOC_current(float Id_des, float Iq_des, float I_phase, float pwm_pha
-    float ilim = effective_current_lim();
+    float ilim = effective_torque_lim() / config_.torque_constant;
     float id = std::clamp(current_control_.Id_setpoint, -ilim, ilim);
     float iq = std::clamp(current_setpoint, -ilim, ilim);
```
Note that `iq = torque_setpoint / torque_constant` doesn't work for ACIM motors. In that case you need to do:
`iq = torque_setpoint / (torque_constant * acim_rotor_flux)`. Beware of dividing by zero acim_rotor_flux, you can use `max(acim_rotor_flux, acim_gain_min_flux)`

PAJohnson committed:
Made changes reflecting PR comments.
Added torque_ramp_rate to controller config.
Changed INPUT_MODE_CURRENT_RAMP to INPUT_MODE_TORQUE_RAMP for controller input mode enum.
Torque limits and current limits are now observed seperately. Torque limit is in the controller, current limit is in the motor object.
Fixed torque -> current calculation in motor_update to handle ACIM motors.

madcowswe commented on Firmware/MotorControl/motor.hpp:
```diff
@@ -159,6 +162,7 @@ class Motor : public ODriveIntf::MotorIntf {
     } gate_driver_exported_;
     DRV_SPI_8301_Vars_t gate_driver_regs_; //Local view of DRV registers (initialized by DRV8301_setup)
     float thermal_current_lim_ = 10.0f;  //[A]
+    float thermal_torque_lim_ = 10.0f; //[Nm]
```
This variable is not used.

madcowswe commented on Firmware/MotorControl/motor.cpp:
```diff
@@ -440,11 +440,19 @@ bool Motor::FOC_current(float Id_des, float Iq_des, float I_phase, float pwm_pha
-bool Motor::update(float current_setpoint, float phase, float phase_vel) {
-    current_setpoint *= config_.direction;
+bool Motor::update(float torque_setpoint, float phase, float phase_vel) {
+    float current_setpoint;
```
Maybe set to 0 by default to avoid magic smoke in case of a code error in the future.

madcowswe commented on Firmware/MotorControl/motor.hpp:
```diff
@@ -45,12 +45,15 @@ class Motor : public ODriveIntf::MotorIntf {
         float resistance_calib_max_voltage = 2.0f; // [V] - You may need to increase this if this voltage isn't sufficient to drive calibration_current through the motor.
         float phase_inductance = 0.0f;        // to be set by measure_phase_inductance
         float phase_resistance = 0.0f;        // to be set by measure_phase_resistance
+        float torque_constant = 1.0f;         // to be set by user
```
Units: Nm/A for PM motors, Nm/A^2 for induction motors.

madcowswe commented on Firmware/MotorControl/controller.cpp:
```diff
@@ -316,36 +316,35 @@ bool Controller::update(float* current_setpoint_output) {
-            vel_integrator_current_ += ((vel_integrator_gain * gain_scheduling_multiplier) * current_meas_period) * v_err;
+            vel_integrator_torque_ += ((vel_integrator_gain * gain_scheduling_multiplier) * current_meas_period) * v_err;
         }
     }
```
Since this no longer considers all kinds of current/torque limits (just the straight torque one), we are not implementing the anti-windup correctly. We need to consider all saturation effects on the "actuator" i.e. torque production.
I will follow up on Discord with some theory.

PAJohnson committed:
Added Motor::max_available_torque(). Use is to correctly determine torque limit for velocity anti-windup.
Set default value (0.0) for current_setpoint
Units documentation
Removed unused variable thermal_torque_lim_

madcowswe commented on Firmware/MotorControl/motor.cpp:
```diff
@@ -192,6 +192,21 @@ float Motor::effective_current_lim() {
+    //return the maximum available torque for the motor.
+    //Note - for ACIM motors, available torque is allowed to be 0.
+    if (config_.motor_type == Motor::MOTOR_TYPE_ACIM) {
+        float max_torque = effective_current_lim() * config_.torque_constant * current_control_.acim_rotor_flux;
```
While not a desired operating point, it can happen that acim_rotor_flux is negative. In that case we probably want to clamp it to 0 to avoid issues with reporting negative max_torque.

madcowswe commented on Firmware/MotorControl/motor.cpp:
```diff
@@ -192,6 +192,21 @@ float Motor::effective_current_lim() {
 
+float Motor::max_available_torque() {
+    //return the maximum available torque for the motor.
+    //Note - for ACIM motors, available torque is allowed to be 0.
```
By convention we usually put general function documentation right above the function definition rather than right below.

PAJohnson committed:
Clamped max_torque to 0 for ACIM motors in case max_torque happened to be negative.

madcowswe committed:
Cleaner clamping

PAJohnson committed:
Changed vel_gain, vel_integrator_gain and torque_constant to work out of the box with ODrive branded motors.
Removed unused variable torque_lim_margin
Changed default value of torque_lim to +inf

PAJohnson committed:
Changed default values of vel_gain, vel_integrator_gain and torque_constant to work with ODrive branded motors.
Changed units in comments from A to Nm where appropriate.

PAJohnson committed:
Merge branch 'A_to_Nm' of https://github.com/PAJohnson/ODrive into A_to_Nm

PAJohnson committed:
Merge remote-tracking branch 'upstream/devel' into A_to_Nm

PAJohnson committed:
Merge remote-tracking branch 'upstream/devel' into A_to_Nm

PAJohnson committed:
Raised current limit from 10A to 15A for regen test. Previously, a motor overcurrent error was triggered ~50% of the time during the braking test.

PAJohnson committed:
Changed docs to reflect removal of current control mode and addition of torque control mode
Regenerated enums
changed can_test.py to reflect control mode change
modified closed_loop_test.py to pass with A->Nm change.

PAJohnson committed:
Added test for torque limit - TestTorqueLimit()

PAJohnson committed:
Merge branch 'devel' of https://github.com/madcowswe/Odrive into A_to_Nm

PAJohnson committed:
Changed docs to reflect change from CONTROL_MODE_CURRENT_CONTROL to CONTROL_MODE_TORQUE_CONTROL

PAJohnson committed:
Change from CurrentControl in interface yaml to TorqueControl
